### ISSUE TITLE
Enhancement: Support "smart indent" where appropriate

### ISSUE THREAD

**Issue by barrywhart** [alanmcruickshank/sqlfluff#76]
This is something to think about, not a "do exactly this" issue.

This query produces a warning:
```
SELECT
    a
FROM
    t
JOIN
    u
USING
    (user_id,
     list_id)
```

Output:
```
$ sqlfluff lint test.sql
== [test.sql] FAIL
L:   9 | P:   1 | L003 | Indentation length is not a multiple of 4.
```

It might be cool to support something like the Python "hanging indent" described [here](https://www.python.org/dev/peps/pep-0008/#fn-hi). I.e. since the second column is indented to align with the first column name, this should be reported as okay (no warning).

There may be other SQL constructs which could/should support the same behavior.

**Comment by barrywhart** [comment]
Another SQL construct where hanging indent inside parentheses makes sense: Analytic functions, e.g.
```
            ROW_NUMBER() OVER
                (PARTITION BY LOW_business_type, LOW_size_label, LOW_gender_label, LOW_age_label
                 ORDER BY business_type DESC, size_label DESC, gender_label DESC, age_label DESC) AS rownum,
```

**Comment by alanmcruickshank** [comment]
Yep I think this is a great idea. I'd like to have a lot more around proper linting of indentation, where to indent and where not to. I'd steered clear of getting too deep into it so far, because I realise it might be one of the more controversial parts of the linting rules.

I think for the first step of this, I might put together something in the docs around what I think the indentation rules should be that sqlfluff tries to enforce. @barrywhart - would you be happy to review that if I wrote something?

**Comment by barrywhart** [comment]
Sure!

Speaking of controversy, we had been using 2-space indent for our SQL. I switched this particular project to 4 as part of testing and adopting SQLFluff. Nobody complained, but I'd be +1 for making the indentation step size configurable.

**Comment by alanmcruickshank** [comment]
I'll be addressing this as part of #91. Discussion will go there.
